File: daily/2024/2601_Prime_Subtraction_Operation.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    bool is_prime(int n){
        for (int i = 2; i*i <= n; i++){
            if (n % i == 0) return false;
        }
        return true;
    }
    bool primeSubOperation(vector<int>& nums) {
        vector<int> prime;
        for (int i = 2; i <= 1000; i++) if (is_prime(i)) prime.push_back(i);
        int n = nums.size();
        for (int i = n-2; i >= 0; i--){
            if (nums[i] < nums[i+1]) continue;
            int idx = 0;
            while(idx < prime.size() && prime[idx] <= nums[i] - nums[i+1]) idx++;
            if (idx == prime.size() || prime[idx] >= nums[i]) return false;
            nums[i] -= prime[idx];
        }
        // for (int i : nums) cout << i << " ";
        // cout << endl;
        return true;
    }
};

int main(){
    vector<int> nums = {4,9,6,10};
    // vector<int> nums = {6,8,11,12};
    // vector<int> nums = {5,8,3};
    Solution S;
    cout << S.primeSubOperation(nums) << endl;
    return 0;
}

File: daily_python/2601_Prime_Subtraction_Operation.py
import os
import sys
from typing import List
from math import sqrt

class Solution:
    def is_prime(self, n):
        for i in range(2, int(sqrt(n)) + 1):
            if n % i == 0:
                return False
        return True
    
    def primeSubOperation(self, nums: List[int]) -> bool:
        M = max(nums)
        lastest_prime = [0] * (M + 1)
        lastest = -1
        for i in range(M, 1, -1):
            lastest_prime[i] = lastest
            if self.is_prime(i):
                lastest = i
        lastest_prime[0] = 2
        lastest_prime[1] = 2
        n = len(nums)
        for i in range(n - 2, -1, -1):
            d = nums[i] - nums[i+1]
            if d < 0:
                continue
            if lastest_prime[d] == -1:
                return False
            nums[i] -= lastest_prime[d]
            if nums[i] <= 0:
                return False
        return True

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            dataset.append(nums)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nums in dataset:
        results.append(solution.primeSubOperation(nums))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')